src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -324,19 +322,24 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
 
                             if (bestConstructorParameterTypes == null)
                             {
-                                bestConstructorParameterTypes = new HashSet<Type>(
-                                    bestConstructor.GetParameters().Select(p => p.ParameterType));
+                                bestConstructorParameterTypes = new HashSet<Type>();
```

Comment:
Would it be worth initializing HashSet with a capacity of bestConstructor.GetParameters().Length ? It could be faster/use less memory if it is not frequent that there is duplicates (IMHO it is not very frequent with dependency injection to have a constructor with parameters of the same type). If it's a hot path, it could yield small perf improvement. Doesn't have to be in this PR, just starting discussion about this.

Reply:
I considered that,  but when looking at the actual sizes (at least in the default templates), about a quarter involved duplicates, and most were smaller than the prime size that'll be rounded up to,  anyway.  Without concrete data suggesting specifying the capacity would be better on average,  I left it as it was.